### ISSUE TITLE
Error while initializing db with docker-compose

### ISSUE THREAD

**Issue by transistorfet** [mggates39/ChipInventory#13]
I followed the instructions for the docker compose workflow.  I created an `.env` file and changed the passwords to something non-empty, and then ran `docker compose up`.  I can connect to the web interface but it shows:

```
Host '172.18.0.3' is not allowed to connect to this MySQL server
```

The docker compose logs shows that an error occurs when initializing the db
```
mysqldb-1  | 2024-11-29T19:28:13.243708Z 6 [Warning] [MY-010453] [Server] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
mysqldb-1  | 2024-11-29T19:28:14.916160Z 6 [ERROR] [MY-000061] [Server] 1008  Can't drop database 'chip_data'; database doesn't exist.
mysqldb-1  | 2024-11-29T19:28:14.919398Z 0 [ERROR] [MY-013236] [Server] The designated data directory /var/lib/mysql/ is unusable. You can remove all files that the server added to it.
mysqldb-1  | 2024-11-29T19:28:14.919410Z 0 [ERROR] [MY-010119] [Server] Aborting
```

I removed the `DROP` query at the start of the init.sql file, and then got another error:
```
mysqldb-1  | 2024-11-29T19:47:31.332672Z 6 [ERROR] [MY-000061] [Server] 1410  You are not allowed to create a user with GRANT.
```

I then added a create user query on line 748 of the init.sql file `CREATE USER 'chip_app'@'%' IDENTIFIED BY 'test123';` and that seemed to create the db.  I still get this error showing up in the logs:
```
CREATE USER 'chip_app'@'%' IDENTIFIED BY 'test123';
```
but it seems to access the db now.  I'm not sure if it's working correctly or not though

Thanks for the project!

**Comment by mggates39** [comment]
do you mind sending your .env file, with passwords scrubbed of course, and I can try to recreate it on my end.

**Comment by transistorfet** [comment]
It's just the defaults from the readme
```
NODE_LOCAL_PORT=3000
NODE_DOCKER_PORT=3000

MYSQL_LOCAL_PORT=3306
MYSQL_DOCKER_PORT=3306
MYSQL_HOST=mysqldb
MYSQL_USER='chip_app'
MYSQL_PASSWORD='test123'
MYSQL_DATABASE='chip_data'
MYSQL_ROOT_PASSWORD='test123'
```

**Comment by mggates39** [comment]
Using your .env file, with my local instance of MySQL stopped, I get the following on a clean build. and then go to http://localhost:3000 in a browser.

mggates@mggates-Latitude-E7470:~/Dev/ChipInventory$ docker compose up
[+] Running 12/12
 ✔ mysqldb Pulled                                                                                                                                                             171.5s 
   ✔ 49bb46380f8c Pull complete                                                                                                                                               163.5s 
   ✔ aab3066bbf8f Pull complete                                                                                                                                               163.5s 
   ✔ d6eef8c26cf9 Pull complete                                                                                                                                               163.6s 
   ✔ 0e908b1dcba2 Pull complete                                                                                                                                               163.8s 
   ✔ 480c3912a2fd Pull complete                                                                                                                                               163.8s 
   ✔ 89a648ecb3cf Pull complete                                                                                                                                               163.9s 
   ✔ 6313eed00780 Pull complete                                                                                                                                               165.2s 
   ✔ 668fe2d98404 Pull complete                                                                                                                                               165.2s 
   ✔ d3f8a843b813 Pull complete                                                                                                                                               170.4s 
   ✔ c80ab9fc8db5 Pull complete                                                                                                                                               170.4s 
   ✔ 1b8b6b073273 Pull complete                                                                                                                                               170.4s 
[+] Building 2.7s (12/12) FINISHED                                                                                                                                    docker:default
 => [app internal] load build definition from Dockerfile                                                                                                                        0.0s
 => => transferring dockerfile: 262B                                                                                                                                            0.0s
 => [app internal] load metadata for docker.io/library/node:20-alpine                                                                                                           0.4s
 => [app internal] load .dockerignore                                                                                                                                           0.0s
 => => transferring context: 113B                                                                                                                                               0.0s
 => [app 1/6] FROM docker.io/library/node:20-alpine@sha256:b5b9467fe7b33aad47f1ec3f6e0646a658f85f05c18d4243024212a91f3b7554                                        0.0s
 => [app internal] load build context                                                                                                                                           0.4s
 => => transferring context: 758.98kB                                                                                                                                           0.4s
 => CACHED [app 2/6] RUN mkdir -p /home/node/app/node_modules && mkdir -p /home/node/app/upload         0.0s
 => CACHED [app 3/6] WORKDIR /home/node/app                                                                                                              0.0s
 => CACHED [app 4/6] COPY package*.json ./                                                                                                                          0.0s
 => CACHED [app 5/6] RUN npm install                                                                                                                                       0.0s
 => [app 6/6] COPY --chown=node:node . .                                                                                                                                  1.1s
 => [app] exporting to image                                                                                                                                                    0.6s
 => => exporting layers                                                                                                                                                         0.5s
 => => writing image sha256:2f7bd706fad2783e45c713e20929adadc1872c304e3535711cf943d52b8158f0       0.0s
 => => naming to docker.io/library/chipinventory-app                                                                                                     0.0s
 => [app] resolving provenance for metadata file                                                                                                             0.0s
[+] Running 3/2
 ✔ Network nodejs-mysql-network       Created                                                                                                                0.2s 
 ✔ Container chipinventory-mysqldb-1  Created                                                                                                              0.1s 
 ✔ Container chipinventory-app-1      Created                                                                                                                  0.0s 
Attaching to app-1, mysqldb-1
mysqldb-1  | 2024-11-29 21:11:52+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.33-1.el8 started.
mysqldb-1  | 2024-11-29 21:11:53+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mysqldb-1  | 2024-11-29 21:11:53+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.33-1.el8 started.
mysqldb-1  | '/var/lib/mysql/mysql.sock' -> '/var/run/mysqld/mysqld.sock'
mysqldb-1  | 2024-11-29T21:11:53.992398Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
mysqldb-1  | 2024-11-29T21:11:53.996091Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.33) starting as process 1
mysqldb-1  | 2024-11-29T21:11:54.019773Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
mysqldb-1  | 2024-11-29T21:11:54.268545Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
mysqldb-1  | 2024-11-29T21:11:54.647373Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
mysqldb-1  | 2024-11-29T21:11:54.647432Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
mysqldb-1  | 2024-11-29T21:11:54.650203Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users. Consider choosing a different directory.
mysqldb-1  | 2024-11-29T21:11:56.116120Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Bind-address: '::' port: 33060, socket: /var/run/mysqld/mysqlx.sock
mysqldb-1  | 2024-11-29T21:11:56.116227Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.33'  socket: '/var/run/mysqld/mysqld.sock'  port: 3306  MySQL Community Server - GPL.
app-1      | GET / 200 510.077 ms - 4861
app-1      | GET /stylesheets/style.css 304 2.951 ms - -

**Comment by transistorfet** [comment]
Hmm.... I don't see the initialization messages in those logs.  I wonder if it used the existing volume instead of initializing the db from scratch.  I deleted the existing database volume with `docker volume rm chipinventory_mysql-data` and then ran it again, and my logs look like this (note the "Initializing database files", "Database files initialized", and "Starting temporary server" messages):
```
mysqldb-1  | 2024-11-29 23:43:09+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.33-1.el8 started.
mysqldb-1  | 2024-11-29 23:43:09+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mysqldb-1  | 2024-11-29 23:43:09+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.33-1.el8 started.
mysqldb-1  | 2024-11-29 23:43:09+00:00 [Note] [Entrypoint]: Initializing database files
mysqldb-1  | 2024-11-29T23:43:09.766156Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
mysqldb-1  | 2024-11-29T23:43:09.766286Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.33) initializing of server in progress as process 80
mysqldb-1  | 2024-11-29T23:43:09.778600Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
mysqldb-1  | 2024-11-29T23:43:10.413652Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
mysqldb-1  | 2024-11-29T23:43:11.980585Z 6 [Warning] [MY-010453] [Server] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
mysqldb-1  | 2024-11-29 23:43:17+00:00 [Note] [Entrypoint]: Database files initialized
mysqldb-1  | 2024-11-29 23:43:17+00:00 [Note] [Entrypoint]: Starting temporary server
mysqldb-1  | 2024-11-29T23:43:18.157661Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
mysqldb-1  | 2024-11-29T23:43:18.159198Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.33) starting as process 136
mysqldb-1  | 2024-11-29T23:43:18.189287Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
mysqldb-1  | 2024-11-29T23:43:18.334076Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
mysqldb-1  | 2024-11-29T23:43:18.553238Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
mysqldb-1  | 2024-11-29T23:43:18.553265Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
mysqldb-1  | 2024-11-29T23:43:18.558357Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users. Consider choosing a different directory.
mysqldb-1  | 2024-11-29T23:43:20.681767Z 7 [ERROR] [MY-000061] [Server] 1396  Operation CREATE USER failed for 'chip_app'@'%'.
mysqldb-1  | 2024-11-29T23:43:20.682734Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Socket: /var/run/mysqld/mysqlx.sock
mysqldb-1  | 2024-11-29T23:43:20.682755Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.33'  socket: '/var/run/mysqld/mysqld.sock'  port: 0  MySQL Community Server - GPL.
mysqldb-1  | 2024-11-29 23:43:20+00:00 [Note] [Entrypoint]: Temporary server started.
mysqldb-1  | '/var/lib/mysql/mysql.sock' -> '/var/run/mysqld/mysqld.sock'
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/iso3166.tab' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/leap-seconds.list' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/leapseconds' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/tzdata.zi' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/zone.tab' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/zone1970.tab' as time zone. Skipping it.
mysqldb-1  | 2024-11-29 23:43:21+00:00 [Note] [Entrypoint]: Creating database chip_data
mysqldb-1  | 2024-11-29 23:43:21+00:00 [Note] [Entrypoint]: Creating user chip_app
mysqldb-1  | ERROR 1396 (HY000) at line 1: Operation CREATE USER failed for 'chip_app'@'%'
```
This is with changes to init.sql for `DROP DATABASE IF EXISTS` , and I added a `CREATE USER` at the bottom just before the `GRANT` but that's now failing when it previously worked.

**Comment by mggates39** [comment]
Looks like you are getting closer.  That log shows that the database created the user in the first steps before the init.sql script runs.  You  should not need the CREATE USER statement in init.sql.

I would drop the volume again,  reset the init.sql script. and try it again.

**Comment by transistorfet** [comment]
Without the `CREATE_USER` statement, it fails with
```
mysqldb-1  | 2024-11-29T23:42:51.619324Z 6 [ERROR] [MY-000061] [Server] 1410  You are not allowed to create a user with GRANT.
```

But if I replace it with `CREATE USER IF NOT EXISTS` it still gives the error about operation failed to create user, but also references the `CREATE USER IF NOT EXISTS` statement, as if it's running the db initialization script multiple times
```
mysqldb-1  | 2024-11-30 00:33:21+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.33-1.el8 started.
mysqldb-1  | 2024-11-30 00:33:21+00:00 [Note] [Entrypoint]: Switching to dedicated user 'mysql'
mysqldb-1  | 2024-11-30 00:33:21+00:00 [Note] [Entrypoint]: Entrypoint script for MySQL Server 8.0.33-1.el8 started.
mysqldb-1  | 2024-11-30 00:33:21+00:00 [Note] [Entrypoint]: Initializing database files
mysqldb-1  | 2024-11-30T00:33:21.763453Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
mysqldb-1  | 2024-11-30T00:33:21.763543Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.33) initializing of server in progress as process 80
mysqldb-1  | 2024-11-30T00:33:21.771483Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
mysqldb-1  | 2024-11-30T00:33:22.315116Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
mysqldb-1  | 2024-11-30T00:33:23.998555Z 6 [Warning] [MY-010453] [Server] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
mysqldb-1  | 2024-11-30 00:33:29+00:00 [Note] [Entrypoint]: Database files initialized
mysqldb-1  | 2024-11-30 00:33:29+00:00 [Note] [Entrypoint]: Starting temporary server
mysqldb-1  | 2024-11-30T00:33:30.014247Z 0 [Warning] [MY-011068] [Server] The syntax '--skip-host-cache' is deprecated and will be removed in a future release. Please use SET GLOBAL host_cache_size=0 instead.
mysqldb-1  | 2024-11-30T00:33:30.015772Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.33) starting as process 136
mysqldb-1  | 2024-11-30T00:33:30.053516Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
mysqldb-1  | 2024-11-30T00:33:30.197312Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
mysqldb-1  | 2024-11-30T00:33:30.458782Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
mysqldb-1  | 2024-11-30T00:33:30.458810Z 0 [System] [MY-013602] [Server] Channel mysql_main configured to support TLS. Encrypted connections are now supported for this channel.
mysqldb-1  | 2024-11-30T00:33:30.462084Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users. Consider choosing a different directory.
mysqldb-1  | 2024-11-30T00:33:32.457374Z 7 [Warning] [MY-010235] [Server] Following users were specified in CREATE USER IF NOT EXISTS but they already exist. Corresponding entry in binary log used default authentication plugin 'caching_sha2_password' to rewrite authentication information (if any) for them: 'chip_app'@'%'
mysqldb-1  | 2024-11-30T00:33:32.477683Z 0 [System] [MY-011323] [Server] X Plugin ready for connections. Socket: /var/run/mysqld/mysqlx.sock
mysqldb-1  | 2024-11-30T00:33:32.477725Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.33'  socket: '/var/run/mysqld/mysqld.sock'  port: 0  MySQL Community Server - GPL.
mysqldb-1  | 2024-11-30 00:33:32+00:00 [Note] [Entrypoint]: Temporary server started.
mysqldb-1  | '/var/lib/mysql/mysql.sock' -> '/var/run/mysqld/mysqld.sock'
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/iso3166.tab' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/leap-seconds.list' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/leapseconds' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/tzdata.zi' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/zone.tab' as time zone. Skipping it.
mysqldb-1  | Warning: Unable to load '/usr/share/zoneinfo/zone1970.tab' as time zone. Skipping it.
mysqldb-1  | 2024-11-30 00:33:33+00:00 [Note] [Entrypoint]: Creating database chip_data
mysqldb-1  | 2024-11-30 00:33:33+00:00 [Note] [Entrypoint]: Creating user chip_app
mysqldb-1  | ERROR 1396 (HY000) at line 1: Operation CREATE USER failed for 'chip_app'@'%'
mysqldb-1 exited with code 1
```

**Comment by mggates39** [comment]
Okay, my docker-fuu is not strong but what I would do is something like this
docker compose down
docker images
docker rmi xxxxxxxx
where xxxxx is the image id of the mysql service container.
then docker compose up

you know, turn it off, nuke it from orbit and turn it back on again.

**Comment by transistorfet** [comment]
Heheh I tried that a couple hours ago. I down'd everything, deleted the mysql image, deleted the volume from the docker container, and ran docker compose up --build, and have still been getting these problems

**Comment by mggates39** [comment]
Then there must be something different between our environments.   Will you have some time tomorrow to get on a discord chat so we can compare notes directly?

**Comment by transistorfet** [comment]
Oh I might have found the problem.  It seemed to be running the init.sql file multiple times from the logs when I added the `CREATE USER IF NOT EXISTS` line, which was a clue.  I looked up `docker mysql init script` and found https://gist.github.com/iamvickyav/86472294dad254052e5c8cac93d69b05 which shows an autoloading directory in the volumes mapped to the container, `/docker-entrypoint-initdb.d/` with a different `command: ` field.  I tried mapping the init script to that autoload dir and commenting out the `command: ` field, and it seems to have worked.  I deleted the volume with `docker volume rm chipinventory_mysql-data`, deleted the containers, and rebuilt it, and I can connect without errors and can view the inventory list.  So I'm pretty sure it did actually load the schema correctly.  It works with the original `init.sql` file without modifications, but with the `docker-compose.yaml` changes
